lib/Zonemaster/Backend/Validator.pm

```diff
@@ -0,0 +1,47 @@
+package Zonemaster::Backend::Validator;
+
+our $VERSION = '0.1.0';
+
+use strict;
+use warnings;
+use 5.14.2;
+
+use JSON::Validator "joi";
+
+
+my $ipv4_regex = "^[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\$";
+my $ipv6_regex = "^([0-9A-Fa-f]{1,4}:[0-9A-Fa-f:]{1,}(:[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3})?)\$|([0-9A-Fa-f]{1,4}::[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3})\$";
+
+our $api_key = joi->string;
+our $batch_id = joi->integer->positive;
+our $client_id = joi->string;
+our $client_version = joi->string;
+our $domain_name = joi->string->max(254);
```

Comment:
254 if ending with a dot, else 253.

$domain_name is the string that comes in, before any conversion from U-label to A-label in case of IDN, isn't it? The length restriction of the A-label or ASCII domain. On the other hand, if the string is too long Engine (test case Basic00) will capture the error.

The API will accept any characters in the domain name, but will reject the any name that can neither be converted to a valid A-label nor matching m/^[\-a-zA-Z0-9\.\_]+$/. We should watch out how we log or report domain names with illegal characters. Maybe we should convert them to strings of Unicode codes and report and log those.

Reply:
@matsduf Yes, it's before any manipulation.
Have you a regex already done for this validation ?

Reply:
The right thing to do for this PR is probably to update the API documentation to match the current implementation, update the validation schema to match that as well. I take upon myself to create an issue about updating both the documentation and the implementation to actually do the right thing.

Current state of normalization:
 1. Alternative label separators ARE NOT replaced with full stop (dot, ".").
 2. Trailing dot IS NOT appended if it's missing.
 3. U-labels ARE encoded into A-labels.
 4. Upper case characters in ASCII labels ARE NOT converted into lower case.

Current state of validation:
 1. Maximum domain name length after A-label encoding IS checked. (The 253/254 distinction is correctly considered)
 2. Maximum label length after A-label encoding IS checked.
 3. Empty domain name IS forbidden. (Although that gives a bogus error message)
 4. The requirement for single-character domain being "." IS NOT checked.
 5. IDNA validity of code points in U-labels before A-label encoding IS NOT checked.
 7. IDNA validity of A-labels IS NOT checked.
 8. IDNA validity of encoded code points in A-labels IS NOT checked.
 9. LDH+underscore validity of octets/characters in ASCII labels IS checked.

Reply:
So, i keep this implementation like that until a new update with a regex or something else. Anyway, i'm not sure that it's possible to implement this kind of validation with the JSON::Validator module.